---
title: "PRJEB23709_PD1_ipi_PD1.2020.03.21 WGCNA"
author: "jianguo.zhou@yahoo.com"
date: "2020.04.25"
output: html_document
---
微信ID: epigenomics  E-mail: epigenomics@rainbow-genome.com

本代码作者：TSS，他来自生信菜鸟团

[<img src="http://p3gnrt3qv.bkt.clouddn.com/bolg/180622/9h1c1i77J2.png?imageslim" width="100px"/>](http://p3gnrt3qv.bkt.clouddn.com/bolg/180622/9h1c1i77J2.png?imageslim)

小丫编译用法、校验代码

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### 需求描述
用R代码画出paper里的WGCNA结果图、同一模块内部的基因表达谱拟合，输出模块基因表达谱，用于绘制heatmap和功能富集分析。

例子一：这篇paper里的WGCNA结果图

![](demo1.jpg)

出自<https://doi.org/10.18632/oncotarget.15842>

例子二：用WGCNA找出gene module，用拟合的方式，展示每个module里的基因表达谱。

![](demo2.png)

出自<http://www.pnas.org/content/112/43/13360>

例子三：用WGCNA找出的module画热图。把各个module的基因作为文件输出，将用于代码9或代码17的输入文件。

![](demo3.png)

出自<https://www.cell.com/cell/abstract/S0092-8674(18)30511-7>

### 应用场景

不需要筛差异表达基因，直接从全基因组找表达谱相似的gene组成的gene module（模块）。尤其适合样本量超过15的高通量测序数据。

用于研究表型的差异是由哪些gene决定的，例如找出决定不同发育阶段、疾病亚型、药物处理反应的关键基因module，进而鉴定新的biomarkers或治疗靶标。

借鉴文章里的应用方案：

#### 方案一：找gene module作为biomarker

例子一的paper用TCGA的papillary renal cell carcinoma (PRCC) RNA-seq数据找出差异表达基因，用WGCNA找出11个共表达的基因模块。其中蓝色模块与病理状态高度相关（r=0.45），其功能富集在nuclear division, cell cycle phase, and spindle。蓝色模块中的40个关键基因可以区分癌症亚型，有望用做biomarker。

#### 方案一：用拟合的曲线展示各gene module的基因表达谱

例子二的paper，用每个module排名前30基因的均值拟合表达谱曲线。

#### 方案二：用热图和表格展示gene module-表达谱-功能-表型的关系

例子三左边的热图是用module里的gene画的，每个module就相当于大家熟悉的热图里的cluster，因此，不用再做聚类。

右侧的富集分析表格的做法：先为每个module里的基因做功能富集分析；然后取排名第一的通路的term，粘贴到该module热图的右侧。


### 输入数据的预处理

输入数据是12219个TPM > 1的基因在20个sample里的TPM。包括5种细胞、2种处理、2次重复。

数据来源：<https://www.cell.com/cms/attachment/2119342979/2092476184/mmc2.xlsx>
TPM参考：<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6421058/>

```{r}
# mydata= read.csv("easy_input_exp.csv",row.names = 1)
# dim(mydata)
# #此处用head(mydata)看出行名是基因名，第1列至第20列是基因在20个sample里的表达量
# head(mydata)
# #如果你的sample比较多，几十上百个的，推荐用fix函数来查看，MAC系统可能要另外安装XQuartz。运行下面这行来代替head(mydata)
# #fix(mydata)
# rm(list=ls())
# 1.1 load data
datExpr=train.expr
```

转置，也就是行变成列，列变成行

```{r}
datExpr0 = data.frame(t(datExpr))
colnames(datExpr0) <- rownames(datExpr)
rownames(datExpr0) <- colnames(datExpr)
```

这套数据的作者用的是TPM > 1的所有基因，如果你也想这样做，就先运行下面这行代码，跳过“筛选方差前25%的基因”，直接进入下一部分。

```{r}
# don't run
datExpr1<-datExpr0[datExpr0>1]
datExpr1<- data.frame(datExpr1)
```

#### 筛选方差前25%的基因

输入数据的作者选取的是TPM>1的基因作为WGCNA的输入。基因数量较多，对电脑要求高。为了便于大家在自己的电脑上操作本套代码，此处挑选了变化较大的前25%的基因。

用前25%的基因会比用全部基因找出来的module数量少。

```{r}
m.vars=apply(datExpr0,2,var)
expro.upper=datExpr0[,which(m.vars>quantile(m.vars, probs = seq(0, 1, 0.25))[4])]
#datExpr1<-data.matrix(expr.mad)
datExpr1<-data.matrix(expro.upper)
#datExpr1<-data.matrix(datExpr0)

```

#### 判断是否有不好的sample或gene

先检查是否有哪个sample或基因表达量缺失，定义成不好的sample或gene

```{r,warning=FALSE,message=FALSE}
library(WGCNA)
gsg = goodSamplesGenes(datExpr1, verbose = 3)
gsg$allOK
```

如果这里返回的结果是`TRUE`，说明所有基因都通过了检查。

如果你用全部基因作为输入，很有可能返回`FALSE`，说明存在不好的基因或sample。

下面的代码就会去除那些不好的基因或sample。

#### 去除不好的sample或gene

```{r}
if (!gsg$allOK){
  # Optionally, print the gene and sample names that were removed:
  if (sum(!gsg$goodGenes)>0) 
     printFlush(paste("Removing genes:", paste(names(datExpr0)[!gsg$goodGenes], collapse = ", ")));
  if (sum(!gsg$goodSamples)>0) 
     printFlush(paste("Removing samples:", paste(rownames(datExpr1)[!gsg$goodSamples], collapse = ", ")));
  # Remove the offending genes and samples from the data:
  datExpr1 = datExpr1[gsg$goodSamples, gsg$goodGenes]
}
```

#### 判断是否有离群样本

通过聚类分析，能看出是否有个别sample跟多数sample的距离较远，决定是否需要去除离群样本。

```{r,fig.width=10, fig.height=5}
sampleTree = hclust(dist(datExpr1), method = "average")
par(cex = 0.7);
par(mar = c(0,4,2,0))
pdf("sampleTree.pdf",width = 20, height = 8)
plot(sampleTree, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5,
     cex.axis = 1.5, cex.main = 2)
```


如果你的数据有离群样本需要去除，就运行下面这段代码。

#### 去除离群样本

例如：以20000作为cutoff，就会去除最左边的3个sample，只剩下89个sample。
 ERR2208907;ERR2208971；ERR2208889
```{r,fig.width=10,fig.height=5}
pdf("Sample clustering to detect outliers.pdf",width = 20, height = 8)
plot(sampleTree, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5, cex.axis = 1.5, cex.main = 2) +
#想用哪里切，就把“h = 35000”和“cutHeight = 35000”中的“500”换成你的cutoff
abline(h = 70, col = "red") #这里按照6e+06，去除了2个离群样本。
clust = cutreeStatic(sampleTree, cutHeight =70, minSize = 10)
keepSamples = (clust==1)
datExpr = datExpr1[keepSamples, ]
nGenes = ncol(datExpr)
nSamples = nrow(datExpr)
dim(datExpr)
```

这套输入数据不需要去除离群样本过,不需要“去除离群样本”。

运行下面的代码，用全部20个样本进行后续的分析：

```{r, fig.width=10, fig.height=6}
datExpr = as.data.frame(datExpr1)
nGenes = ncol(datExpr)
nSamples = nrow(datExpr)
```

### 找gene module

WGCNA分析的关键是找gene module。先选择合适的阈值，通过构建网络找gene module，找出来的gene module可信度如何？要做Preservation，去除not preserved module。这样找出的共表达的gene module就可以用于下一步分析了。

#### 选择构建网络的合适阈值

通过这步计算，找出scale free topology modle fit接近0.9的最小power（soft threshold），用于下一步构建网络。

```{r, warning=FALSE, message=FALSE}
powers = c(c(1:10), seq(from = 12, to=20, by=2))

sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)

pdf("1Threshold.pdf",width = 10, height = 5)
par(mfrow = c(1,2))
cex1 = 0.9
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
     main = paste("Scale independence")) +
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,cex=cex1,col="red")+
abline(h=0.90,col="red")

plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
     main = paste("Mean connectivity")) +
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
dev.off()
```

![](1Threshold.pdf)

从上面的结果可以看出，从7开始进入“平台期”。因此，我们把下面代码里的`power`设置为`power =14`

#### 构建网络，找出gene module

```{r}
datExpr_n= datExpr
for (i in 1:nrow(datExpr)) {
 datExpr[i,]= as.numeric(datExpr_n[i,]) 
}
net = blockwiseModules(datExpr, power = 7,
                       TOMType = "unsigned", minModuleSize = 30,
                       reassignThreshold = 0, mergeCutHeight = 0.25,
                       numericLabels = TRUE, pamRespectsDendro = FALSE,
                       saveTOMs = TRUE,
                       #saveTOMFileBase = "MyTOM",
                       verbose = 3)

table(net$colors)
```

一共找出13个module，第二行分别对应每个module内的基因个数。

#### gene module的可视化

```{r}
mergedColors = labels2colors(net$colors)

pdf("2module.pdf",width = 10, height = 5)
plotDendroAndColors(net$dendrograms[[1]], mergedColors[net$blockGenes[[1]]], "Module colors",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
dev.off()
moduleLabels = net$colors
moduleColors = labels2colors(net$colors)
MEs = net$MEs;
geneTree = net$dendrograms[[1]]
```

![](2module.pdf)

#### 聚类信息和性状的可视化
```{r}
#明确样本数和基因数
nGenes = ncol(datExpr)
nSamples = nrow(datExpr)
# samples1= GSE33774.clin
# samples1_n= samples1
# for (i in 1:nrow(datExpr)) {
#  samples1[i,]= as.numeric(samples1_n[i,]) 
# }
samples1<-train.clin
#首先针对样本做个系统聚类树
datExpr_tree<-hclust(dist(datExpr), method = "average")
par(mar = c(0,5,2,0))
plot(datExpr_tree, main = "Sample clustering", sub="", xlab="", cex.lab = 2, 
     cex.axis = 1, cex.main = 1,cex.lab=1)
## 如果这个时候样本是有性状，或者临床表型的，可以加进去看看是否聚类合理
#针对前面构造的样品矩阵添加对应颜色
# sample_colors <- numbers2colors(as.numeric(factor(datTraits$Tumor.Type)), 
#                                 colors = c("white","blue","red","green"),signed = FALSE)
## 这个给样品添加对应颜色的代码需要自行修改以适应自己的数据分析项目。
sample_colors <- numbers2colors( samples1,signed = FALSE)
rownames(sample_colors)=rownames(samples1)
colnames(sample_colors)=colnames(samples1)
## 如果样品有多种分类情况，而且 datTraits 里面都是分类信息，那么可以直接用上面代码，当然，这样给的颜色不明显，意义不大。
#构造10个样品的系统聚类树及性状热图
par(mar = c(1,4,3,1),cex=0.8)
plotDendroAndColors(datExpr_tree, sample_colors,
                    groupLabels = colnames(sample),
                    cex.dendroLabels = 0.8,
                    marAll = c(1, 4, 3, 1),
                    cex.rowText = 0.01,
                    main = "Sample dendrogram and trait heatmap")
                    
```               

#### 绘制模块之间相关性图
```{r}
# module eigengene, 可以绘制线图，作为每个模块的基因表达趋势的展示
MEs = net$MEs

### 不需要重新计算，改下列名字就好
### 官方教程是重新计算的，起始可以不用这么麻烦
MEs_col = MEs
library("stringr")
colnames(MEs_col) = paste0("ME", labels2colors(
  as.numeric(str_replace_all(colnames(MEs),"ME",""))))
MEs_col = orderMEs(MEs_col)
```

```{r}
# 根据基因间表达量进行聚类所得到的各模块间的相关性图
# marDendro/marHeatmap 设置下、左、上、右的边距
pdf("2.1plotEigengeneNetworks.pdf",width = 10, height = 5)
plotEigengeneNetworks(MEs_col, "Eigengene adjacency heatmap", 
                      marDendro = c(3,3,2,4),
                      marHeatmap = c(3,4,2,2), plotDendrograms = T, 
                      xLabelsAngle = 90)
dev.off()

```

```{r}
## 如果有表型数据，也可以跟ME数据放一起，一起出图
pdf("2.2plotEigengeneNetworks.samples.pdf",width = 10, height = 7)
MEs_colpheno = orderMEs(cbind(MEs_col, samples1))
plotEigengeneNetworks(MEs_colpheno, "Eigengene adjacency heatmap", 
                      marDendro = c(3,3,2,4),
                      marHeatmap = c(3,4,2,2), plotDendrograms = T, 
                      xLabelsAngle = 90)
dev.off()

```
 
 #### 可视化基因网络 (TOM plot)
```{r}
# 如果采用分步计算，或设置的blocksize>=总基因数，直接load计算好的TOM结果
# 否则需要再计算一遍，比较耗费时间
# TOM = TOMsimilarityFromExpr(dataExpr, power=power, corType=corType, networkType=type)
load(net$TOMFiles[1], verbose=T)

## Loading objects:
##   TOM

TOM <- as.matrix(TOM)
save(TOM,file = "TOM.Rdata")
dissTOM = 1-TOM
# Transform dissTOM with a power to make moderately strong 
# connections more visible in the heatmap
plotTOM = dissTOM^7
# Set diagonal to NA for a nicer plot
diag(plotTOM) = NA
# Call the plot function

# 这一部分特别耗时，行列同时做层级聚类
# TOMplot(plotTOM, net$dendrograms, moduleColors,
#         main = "Network heatmap plot, all genes")
        
```

这个非常消耗计算资源和时间，所以建议选取其中部分基因作图即可，我就没有画，而且根据下面的代码选取部分基因来作图！

#### 然后随机选取部分基因作图
```{r}
nSelect = 400
# For reproducibility, we set the random seed
set.seed(10);
select = sample(nGenes, size = nSelect);
selectTOM = dissTOM[select,select];
# selectTOM<-dissTOM
# There’s no simple way of restricting a clustering tree to a subset of genes, so we must re-cluster.
selectTree = hclust(as.dist(selectTOM), method = "average")
selectColors = moduleColors[select];
#selectColors = moduleColors;

# Open a graphical window
sizeGrWindow(9,9)
# Taking the dissimilarity to a power, say 10, makes the plot more informative by effectively changing
# the color palette; setting the diagonal to NA also improves the clarity of the plot
plotDiss = selectTOM^7;
diag(plotDiss) = NA;
pdf("2.3TOMplot.random.pdf",width = 10, height = 7)
TOMplot(plotDiss, selectTree, selectColors, main = "Network heatmap plot, selected genes")
dev.off()

```


#### module的可重复性reproducible (or preserved)

上面找出来的module是否真的存在？它是否会随机出现？需要通过模块的preservation分析，证实module的可重复性。
ENSG00000251562.8 MALAT1
由于做preservation分析需要两个表达矩阵，即训练和测试，这里我随机切分之前的表达矩阵，来对训练数据集做保守性分析，当然最好一开始就做切分，我这里直接再次切分，给大家提供参考。

##### 1. 建立训练集和测试集

```{r}
# library(caret)
# load("~/jianguozhou/PRJEB23709_WGCNA/PRJEB23709_PD1_ipi_PD1.2020.03.21.Rdata")
# clin=PRJEB23709_PD1_ipi_PD1
# # ERR2208907;ERR2208971；ERR2208889
# clin=clin[clin$Run_accession %in% row.names(datExpr),]
# inTraining <- createDataPartition(clin$Treatment, p = 0.6, list = FALSE)
# train_id<- clin[inTraining,]
# test_id<-clin[-inTraining,]
# train<- data.frame(datExpr[row.names(datExpr) %in% train_id$Run_accession, ]) 
# test<- data.frame(datExpr[row.names(datExpr) %in% test_id$Run_accession, ])
# 
#   b <- apply(train,2,median)
#   xx<- train[,-which(b==0)]
#   train_1=data.frame(xx)
#   
#    b <- apply(test,2,median)
#     xx<- test[,-which(b==0)]
#   test_1=data.frame(xx)
#   test=test_1[,colnames(test_1) %in% colnames(train_1)]
#   table(colnames(test_1) %in% colnames(train_1))
#   train=train_1[,colnames(train_1) %in%  colnames(test_1) ]
#   #在筛选 然后重新跑
# datExpr=datExpr[,colnames(datExpr)%in%  colnames(test) ]

etLabels = c("Train", "Test");
train = datExpr
test1 = test.expr[rownames(test.expr) %in% colnames(datExpr),]
test = data.frame(t(test1))
colnames(test) <- rownames(test1)
rownames(test) <- colnames(test1)
test_n= test
for (i in 1:nrow(test)) {
 test[i,]= as.numeric(test_n[i,]) 
}
rm(multiExpr)
multiExpr = list(Train = list(data = train), Test = list(data = test));
multiColor = list(Train = moduleColors);
```

##### 2. preservation分析

此处为了让大家快点跑通代码，把nPermutations设置为20，这会导致每次生成的module的Z值不一样。官方推荐`nPermutations = 200`

```{r}
nSets = 2
mp = modulePreservation(multiExpr, multiColor,
                        referenceNetworks = 1,
                        nPermutations = 200 ,
                        randomSeed = 1,
                        quickCor = 0,
                        verbose = 3)
save(mp,file = "mp.Rda") #这个很费时间，建议保存一下
```

**推荐的阈值**

Z大于10，代表strong preserved，好的module

大于2小于10代表weak preserved

小于2代表not preserved，不好的module

**参考文献**

Langfelder P, Luo R, Oldham MC, Horvath S (2011) **Is my network module preserved and reproducible?** PloS Comp Biol. 7(1): e1001057 <http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1001057>

```{r}
ref = 1
test = 2
statsObs = cbind(mp$quality$observed[[ref]][[test]][, -1], mp$preservation$observed[[ref]][[test]][, -1])
statsZ = cbind(mp$quality$Z[[ref]][[test]][, -1], mp$preservation$Z[[ref]][[test]][, -1]);
print(cbind(statsObs[, c("medianRank.pres", "medianRank.qual")],
             signif(statsZ[, c("Zsummary.pres", "Zsummary.qual")], 2)) )

modColors = rownames(mp$preservation$observed[[ref]][[test]])
moduleSizes = mp$preservation$Z[[ref]][[test]][, 1]

#这几个module的Z低于10
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4158056/
row.names(statsZ[statsZ$Zsummary.pres<10,])

#去掉Z<10的module
plotMods = !(modColors %in% row.names(statsZ[statsZ$Zsummary.pres<10,]))
text = modColors[plotMods];
plotData = cbind(mp$preservation$observed[[ref]][[test]][, 2], mp$preservation$Z[[ref]][[test]][, 2])
```

##### preservation的可视化

```{r}
mains = c("Preservation Median rank", "Preservation Zsummary");

sizeGrWindow(10, 10);
pdf("3preservation.pdf",width = 20, height = 10)
par(mfrow = c(1,2))
par(mar = c(4.5,4.5,2.5,1))
for (p in 1:2){
  min = min(plotData[, p], na.rm = TRUE);
  max = max(plotData[, p], na.rm = TRUE);
  # Adjust ploting ranges appropriately
  if (p==2){
    if (min > -max/10) min = -max/10
    ylim = c(min - 0.1 * (max-min), max + 0.1 * (max-min))
  } else
    ylim = c(min - 0.1 * (max-min), max + 0.1 * (max-min))
  plot(moduleSizes[plotMods], plotData[plotMods, p], col = 1, bg = modColors[plotMods], pch = 21,
       main = mains[p],
       cex = 2.4,
       ylab = mains[p], xlab = "Module size", log = "x",
       ylim = ylim,
       xlim = c(10, 2000), cex.lab = 1.2, cex.axis = 1.2, cex.main =1.4)
  labelPoints(moduleSizes[plotMods], plotData[plotMods, p], text, cex = 1, offs = 0.08);
  # For Zsummary, add threshold lines
  if (p==2){
    abline(h=0)
    abline(h=2, col = "blue", lty = 2)
    abline(h=5, col = "darkgreen", lty = 2)
  }
}
dev.off()
```

![](3preservation.pdf)

#### 把gene module输出到文件

输出Z>10的每个module内的基因，用于后续分析作图，随便你DIY

```{r}
for (i  in 1:length(text)) {
  y=t(assign(paste(text[i],"test.expr",sep = "."),datExpr[,moduleColors==text[i]]))
  write.csv(y,paste(text[i],"csv",sep = "."),quote = F)
}

```

```{r}
for (i  in 1:length(text)) {
  y=t(assign(paste(text[i],".expr",sep = "."),test_n[,moduleColors==text[i]]))
  write.csv(y,paste(text[i],"test.csv",sep = "."),quote = F)
}
```

### gene module的深入分析

#### 表型与模块的相关性

通过计算表型与module的相关性，找出相关性高的gene module，推测可能是因为它们造成了表型差异。

```{r}
library(readr)
# samples <- clin
# samples=data.frame(samples)
# row.names(samples)=samples$Run_accession
# samples1=samples[,c(2,3,4,5,6,7,11,12,13,19)]
# write.csv(samples1,file = "samples1.csv")
# ##处理后在上传
# library(readr)
# samples1 <- data.frame(read_csv("samples1.csv")) 
# row.names(samples1)=samples1$X1
# samples1=samples1[,-1]
# save(samples1,file = "samples1.Rdata")
moduleLabelsAutomatic = net$colors
moduleColorsAutomatic = labels2colors(moduleLabelsAutomatic)
moduleColorsWW = moduleColorsAutomatic
MEs0 = moduleEigengenes(datExpr, moduleColorsWW)$eigengenes
MEsWW = orderMEs(MEs0)
modTraitCor = cor(MEsWW, samples1, use = "p")
colnames(MEsWW)
modlues=MEsWW

modTraitP = corPvalueStudent(modTraitCor, nSamples)
textMatrix = paste(signif(modTraitCor, 2), "\n(", signif(modTraitP, 1), ")", sep = "")
dim(textMatrix) = dim(modTraitCor)
```

展示全部module跟表型之间的相关性

```{r}
pdf("4Module-trait.pdf",width = 4, height = 10)
labeledHeatmap(Matrix = modTraitCor, xLabels = colnames(samples1), yLabels = names(MEsWW), cex.lab = 0.5,  yColorWidth=0.01, 
               xColorWidth = 0.03,
               ySymbols = colnames(modlues), colorLabels = FALSE, colors = blueWhiteRed(50), 
               textMatrix = textMatrix, setStdMargins = FALSE, cex.text = 0.5, zlim = c(-1,1)
               , main = paste("Module-trait relationships"))
dev.off()
```

![](4Module-trait.pdf)

此处blue的gene module与timepoint_4w的相关性高，而且Z>10，推测blue module里的基因可能对timepoint_4w组的贡献较大。

 TEST 组！！通过计算表型与module的相关性，找出相关性高的gene module，推测可能是因为它们造成了表型差异。

```{r}
moduleLabelsAutomatic = net$colors
moduleColorsAutomatic = labels2colors(moduleLabelsAutomatic)
moduleColorsWW = moduleColorsAutomatic
MEs0 = moduleEigengenes(t(test.expr), moduleColorsWW)$eigengenes
MEsWW.test = orderMEs(MEs0)
modTraitCor.test = cor(MEsWW.test, test.clin, use = "p")
colnames(MEsWW.test)
modlues=MEsWW.test
nSamples.test=18# test 18 pts
modTraitP.test = corPvalueStudent(modTraitCor.test, nSamples.test)
textMatrix.test = paste(signif(modTraitCor.test, 2), "\n(", signif(modTraitP.test, 1), ")", sep = "")
dim(textMatrix.test) = dim(modTraitCor.test)
```

展示全部module跟表型之间的相关性

```{r}
pdf("4Module-trait.test.pdf",width = 4, height = 10)
labeledHeatmap(Matrix = modTraitCor.test, xLabels = colnames(test.clin), yLabels = names(MEsWW.test), cex.lab = 0.5,  yColorWidth=0.01, 
               xColorWidth = 0.03,
               ySymbols = colnames(modlues), colorLabels = FALSE, colors = blueWhiteRed(50), 
               textMatrix = textMatrix.test, setStdMargins = FALSE, cex.text = 0.5, zlim = c(-1,1)
               , main = paste("Module-trait relationships.test"))
dev.off()
```


```{r}
#7.3 hub genes表达热图 
## blue vs age or timepoint_4w
hubexpr<-as.data.frame(blue.expr)
annotation_col <- data.frame(group = samples1$responder, row.names=row.names(hubexpr))
# hubexpr<-hubexpr[,order(dd1_norm["EZH2",],decreasing = T)] #对样品按照EZH2表达水平排序
#mycolors <- colorRampPalette(c("blue","red"), bias = 1.5)(150)
hubexpr<-scale(hubexpr)
hubexpr[hubexpr > 2] = 1
hubexpr[hubexpr < -2] = -1
pheatmap::pheatmap(hubexpr,
                   annotation_col = annotation_col,
                   annotation_legend= TRUE,
                   show_rownames = T, 
                   show_colnames = F,
                   cluster_cols = F, 
                   cluster_rows = T,
                   border_color = "NA")

library(pheatmap)
pheatmap(t(hubexpr), scale = "row", clustering_distance_rows = "correlation")

```

#8 hub genes 功能注释
## 8.1 kegg and go 富集分析

```{r}
if(!require("org.Hs.eg.db")) BiocManager::install("org.Hs.eg.db",ask = F,update = F)  
library(org.Hs.eg.db)
library(clusterProfiler)
modProbes <- bitr(colnames(hubexpr),fromType = "SYMBOL",toType = c("GENENAME","SYMBOL","ENTREZID"),OrgDb = org.Hs.eg.db)
kk <- enrichKEGG(gene         = modProbes$ENTREZID,
                 organism     = 'hsa',
                 qvalueCutoff = 0.1,
                 pvalueCutoff = 0.05,
                 keyType      = "kegg")

ego <- enrichGO(gene          = modProbes$ENTREZID,
                OrgDb         = org.Hs.eg.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.1,
                readable      = TRUE)
head(ego)
head(kk)
```

#8.2 富集结果可视化
#具体参见y叔的enrichplot包说明

```{r}
library(enrichplot)
dotplot(kk)+ggtitle("dotplot for GO enrichement")
cnetplot(kk)
upsetplot(ego)
heatplot(ego)
emapplot(ego)
```


#8.3 One of the problem of enrichment analysis is to find pathways for further investigation. 
#    Here, we provide pmcplot function to plot the number/proportion of publications trend based on 
#    the query result from PubMed Central. 
#    Of course, users can use pmcplot in other scenarios. 
#    All text that can be queried on PMC is valid as input of pmcplot
```{r}
terms <- ego$Description[1:3]
p <- pmcplot(terms, 2010:2017)
p2 <- pmcplot(terms, 2010:2017, proportion=FALSE)
plot_grid(p, p2, ncol=2)

```



#### gene module内基因表达谱的拟合

例子二paper里的数据是时间序列，作者提取表达量最高的30个基因，取平均值（averaged gene expression of top 30）绘制表达谱曲线。

此处取brown的前30个基因，每种细胞取1个样本，假装是5个时间点：

```{r,warning=FALSE}
df30<-read.csv("blue.csv",row.names = 1)
#每种细胞取1个样本，假装是5个时间点
df30time<-df30[,c(1,30)]
head(df30time)

#算30个基因的平均值，取log2
dfave<-data.frame(log2(colMeans(df30time)))
dfave$t<-1:2
colnames(dfave)<-c("avelog2","time")
head(dfave)
```

默认用loess模型拟合，你还可以添加method参数，使用其他方法拟合，例如：`method = "lm"`, 或 "glm", "gam", "loess", "rlm"

```{r, warning=FALSE, message=FALSE,fig.width = 6, fig.height = 3}
library(ggplot2)
ggplot(dfave,aes(x = time,y = avelog2)) +
  geom_smooth(color="black") + #拟合曲线的颜色
  labs(x = "Days after injury", y = "Expression Level", title = "") + 
  theme_bw() + #去除背景色
  theme(panel.grid =element_blank()) + #去除网格线
  theme(panel.border = element_blank()) + #去除外层边框
  theme(axis.line = element_line(colour = "grey")) + #坐标轴画成灰色
  theme(axis.ticks = element_blank()) + #取掉坐标轴上的刻度线
  geom_hline(yintercept = dfave$avelog2[1],linetype="dashed") #在第一个时间点处画虚线
```

```{r}
sessionInfo()
```
